### PR TITLE
fix(float): add fixed option

### LINKED ISSUE #10811: floats: option to truncate width instead of moving on too narrow screen
Currently floating windows are moved left if they would overflow the screen to the right ("do not fit `width`wise").

Vim has a "fixed" option for popup windows that corresponds to this:
```
	fixed		When FALSE (the default), and:
			 - "pos" is "botleft" or "topleft", and
			 - "wrap" is off, and
			 - the popup would be truncated at the right edge of
			   the screen, then
			the popup is moved to the left so as to fit the
			contents on the screen.  Set to TRUE to disable this.
```

Without this you would have to take care of passing in an appropriate `width` (which must take into account any vertical scrolling (sidescroll) also.

I am using `{'relative': 'cursor', 'anchor': 'NW', …}`.

The use case is displaying call signatures (within jedi-vim), which should be positioned at the opening parenthesis.
I am specifying the `width` according to the longest line in there.

While this appears to work it would be nice if it could be avoided with something similar to Vim's "fixed" option:

```vim
cur_col = vim.current.window.cursor[1]
floatwin_col_offset = sig_bracket_start_col - cur_col - 1

" Extra processing to not make Nvim move it left.
start_wincol = vim.funcs.wincol() + floatwin_col_offset
max_width = vim.options["columns"] - start_wincol + 1
width = min([max_sig_width, max_width])

opts = {
    'relative': 'cursor',
    'width': width,
    'height': len(lines),
    'col': floatwin_col_offset,
    'row': -1 * len(lines),
    'anchor': 'NW',
}
win = vim.api.open_win(buf, 0, opts)
```

### PR DESCRIPTION
**Pull request by glepnir**
Problem: We don't have an option (like vim's 'fixed') to truncate float screen when screen width overflows the screen.

Solution: Add `fixed` in `nvim_open_win`.

Fix https://github.com/neovim/neovim/issues/10811

### CODE REVIEW

File: src/nvim/api/win_config.c [hunk 1]
```diff
@@ -160,6 +160,9 @@
 ///   - noautocmd: If true then no buffer-related autocommand events such as
 ///                  |BufEnter|, |BufLeave| or |BufWinEnter| may fire from
 ///                  calling this function.
+///   - fixed: When anchor is NW or SW and wrap is off and the float window would be
```
Comment (bfredl) [inline src/nvim/api/win_config.c:163]: I would rephrase this as putting what the option enables upfront. something like: the position is kept fixed even if the window would be truncated.

File: src/nvim/window.c [hunk 2]
```diff
@@ -1010,6 +1010,11 @@ void ui_ext_win_position(win_T *wp, bool validate)
       comp_col += grid->comp_col;
       comp_row = MAX(MIN(comp_row, Rows - wp->w_height_outer - (p_ch > 0 ? 1 : 0)), 0);
       comp_col = MAX(MIN(comp_col, Columns - wp->w_width_outer), 0);
+      int right_extra = Columns - (int)c.col - wp->w_width - (int)strlen(c.border_chars[2]);
+      if ((c.anchor == 0 || c.anchor == kFloatAnchorSouth)
```
Comment (bfredl) [inline src/nvim/window.c:1014]: could use `!(c.anchor & kFloatAnchorEast)`

File: src/nvim/window.c [hunk 3]
```diff
@@ -1010,6 +1010,10 @@ void ui_ext_win_position(win_T *wp, bool validate)
       comp_col += grid->comp_col;
       comp_row = MAX(MIN(comp_row, Rows - wp->w_height_outer - (p_ch > 0 ? 1 : 0)), 0);
       comp_col = MAX(MIN(comp_col, Columns - wp->w_width_outer), 0);
+      int right_extra = Columns - (int)c.col - wp->w_width - (int)strlen(c.border_chars[2]);
```
Comment (bfredl) [inline src/nvim/window.c:1013]: no strlen, just check whether `c.border_chars[2][0] == NUL)` (width of border is always one)

Comment (glepnir) [reply src/nvim/window.c]: done.

File: src/nvim/window.c [hunk 4]
```diff
@@ -1010,6 +1010,10 @@ void ui_ext_win_position(win_T *wp, bool validate)
       comp_col += grid->comp_col;
       comp_row = MAX(MIN(comp_row, Rows - wp->w_height_outer - (p_ch > 0 ? 1 : 0)), 0);
       comp_col = MAX(MIN(comp_col, Columns - wp->w_width_outer), 0);
+      int right_extra = Columns - (int)c.col - wp->w_width - (c.border_chars[2][0] != 0);
+      if (!(c.anchor & kFloatAnchorEast) && c.fixed && !wp->w_p_wrap && right_extra < 0) {
```
Comment (zeertzjq) [inline src/nvim/window.c:1014]: Why check for `w_p_wrap`? This sounds like something that should be checked by the called that sets `fixed`.

**Comment by zeertzjq** [comment]
Also please fix the typo in the commit message.

File: src/nvim/window.c (hunk 4 above)
Comment (glepnir) [reply src/nvim/window.c:1014]: cause vim help said there check wrap . when wrap on it will not fixed float window width on screen.

```
                         - "pos" is "botleft" or "topleft", and
                         - "wrap" is off, and
                         - the popup would be truncated at the right edge of
                           the screen, then 
```

Comment (zeertzjq) [reply src/nvim/window.c:1014]: This sounds like a completely unnecessary restriction. I don't think it makes sense.

Comment (glepnir) [reply src/nvim/window.c]: removed.

**Comment by glepnir** [comment]
good to go ?

**Comment by zeertzjq** [comment]
You still haven't fixed the typo...
